File: sonar-server/src/main/webapp/WEB-INF/app/models/api/pagination.rb
#
# Sonar, entreprise quality control tool.
# Copyright (C) 2008-2012 SonarSource
# mailto:contact AT sonarsource DOT com
#
# Sonar is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 3 of the License, or (at your option) any later version.
#
# Sonar is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public
# License along with Sonar; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
#
class Api::Pagination

  DEFAULT_PER_PAGE=25
  attr_accessor :per_page, :page, :results

  def initialize(options={})
    @per_page=options[:per_page]||DEFAULT_PER_PAGE
    @page=options[:page].to_i
    @page=1 if @page<1
    @results = options[:results].to_i
  end

  def pages
    @pages ||=
        begin
          p=(@results / @per_page)
          p+=1 if @results % @per_page > 0
          p
        end
  end

  def offset
    (page-1) * per_page
  end

  # alias
  def limit
    per_page
  end

  # inclusive index
  #def to_index
  #  [results-1, (page * per_page)-1].min
  #end

  def previous?
    page>1
  end

  def next?
    page<pages
  end
end

File: server/sonar-db-dao/src/main/java/org/sonar/db/Pagination.java
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.db;

import javax.annotation.concurrent.Immutable;

import static com.google.common.base.Preconditions.checkArgument;

@Immutable
public final class Pagination {
  private static final Pagination ALL = new Builder(1).andSize(Integer.MAX_VALUE);

  private final int page;
  private final int pageSize;

  private Pagination(Builder builder) {
    this.page = builder.page;
    this.pageSize = builder.pageSize;
  }

  public static Pagination all() {
    return ALL;
  }

  public static Builder forPage(int page) {
    return new Builder(page);
  }

  public int getPage() {
    return page;
  }

  public int getPageSize() {
    return pageSize;
  }

  public int getOffset() {
    return (page - 1) * pageSize;
  }

  public int getStartRowNumber() {
    return getOffset() + 1;
  }

  public int getEndRowNumber() {
    return page * pageSize;
  }

  public static final class Builder {
    private final int page;
    private int pageSize = 0;

    public Builder(int page) {
      checkArgument(page >= 1, "page index must be >= 1");
      this.page = page;
    }

    public Pagination andSize(int pageSize) {
      checkArgument(pageSize >= 1, "page size must be >= 1");
      this.pageSize = pageSize;
      return new Pagination(this);
    }
  }
}
